File: modules/virtual-people-core-serving/src/test/kotlin/org/wfanet/virtualpeople/core/model/utils/HashFieldMaskMatcherTest.kt
// Copyright 2022 The Cross-Media Measurement Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package org.wfanet.virtualpeople.core.model.utils

import com.google.protobuf.fieldMask
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertTrue
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.JUnit4
import org.wfanet.virtualpeople.common.labelerEvent

@RunWith(JUnit4::class)
class HashFieldMaskMatcherTest {

  @Test
  fun `empty events should throw`() {
    val hashFieldMask = fieldMask { paths.add("person_country_code") }
    val exception =
      assertFailsWith<IllegalStateException> {
        HashFieldMaskMatcher.build(events = listOf(), hashFieldMask)
      }
    assertTrue(exception.message!!.contains("The events is empty"))
  }

  @Test
  fun `empty hash field mask should throw`() {
    val event = labelerEvent { personCountryCode = "COUNTRY_1" }
    val exception =
      assertFailsWith<IllegalStateException> {
        HashFieldMaskMatcher.build(listOf(event), fieldMask {})
      }
    assertTrue(exception.message!!.contains("The hashFieldMask is empty"))
  }

  @Test
  fun `events have same hash should throw`() {
    val event1 = labelerEvent {
      personCountryCode = "COUNTRY_1"
      personRegionCode = "REGION_1"
    }
    val event2 = labelerEvent {
      personCountryCode = "COUNTRY_1"
      personRegionCode = "REGION_2"
    }
    val hashFieldMask = fieldMask { paths.add("person_country_code") }
    val exception =
      assertFailsWith<IllegalStateException> {
        HashFieldMaskMatcher.build(listOf(event1, event2), hashFieldMask)
      }
    assertTrue(exception.message!!.contains("Multiple events have the same hash"))
  }

  @Test
  fun `getMatch return expected results`() {
    val inputEvent1 = labelerEvent {
      personCountryCode = "COUNTRY_1"
      personRegionCode = "REGION_1"
    }
    val inputEvent2 = labelerEvent {
      personCountryCode = "COUNTRY_2"
      personRegionCode = "REGION_2"
    }
    val hashFieldMask = fieldMask { paths.add("person_country_code") }
    val matcher = HashFieldMaskMatcher.build(listOf(inputEvent1, inputEvent2), hashFieldMask)

    /** events[0] matches. Returns 0. */
    val event1 = labelerEvent { personCountryCode = "COUNTRY_1" }
    assertEquals(0, matcher.getMatch(event1))

    /** events[0] matches. Returns 1. person_region_code is ignored. */
    val event2 = labelerEvent {
      personCountryCode = "COUNTRY_1"
      personRegionCode = "REGION_3"
    }
    assertEquals(0, matcher.getMatch(event2))

    /** events[1] matches. Returns 1. */
    val event3 = labelerEvent { personCountryCode = "COUNTRY_2" }
    assertEquals(1, matcher.getMatch(event3))

    /** events[1] matches. Returns 1. person_region_code is ignored. */
    val event4 = labelerEvent {
      personCountryCode = "COUNTRY_2"
      personRegionCode = "REGION_3"
    }
    assertEquals(1, matcher.getMatch(event4))

    /** No matches. Returns -1. */
    val event5 = labelerEvent { personCountryCode = "COUNTRY_3" }
    assertEquals(-1, matcher.getMatch(event5))
  }

  @Test
  fun `unset field in mask matches event unset`() {
    val inputEvent1 = labelerEvent { personRegionCode = "REGION_1" }
    val inputEvent2 = labelerEvent { personRegionCode = "REGION_2" }
    val hashFieldMask = fieldMask {
      paths.add("person_country_code")
      paths.add("person_region_code")
    }
    val matcher = HashFieldMaskMatcher.build(listOf(inputEvent1, inputEvent2), hashFieldMask)

    /** events[0] matches. Returns 0. */
    val event1 = labelerEvent { personRegionCode = "REGION_1" }
    assertEquals(0, matcher.getMatch(event1))

    /**
     * No matches. Returns -1. When using hash field mask, to match an event with unset field, the
     * event being checked must have the same field unset.
     */
    val event2 = labelerEvent {
      personCountryCode = "COUNTRY_1"
      personRegionCode = "REGION_1"
    }
    assertEquals(-1, matcher.getMatch(event2))
  }
}

File: modules/virtual-people-core-serving/src/test/cc/wfa/virtual_people/core/model/utils/hash_field_mask_matcher_test.cc
// Copyright 2021 The Cross-Media Measurement Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "wfa/virtual_people/core/model/utils/hash_field_mask_matcher.h"

#include <memory>
#include <string>
#include <utility>
#include <vector>

#include "absl/status/status.h"
#include "absl/status/statusor.h"
#include "common_cpp/testing/status_macros.h"
#include "common_cpp/testing/status_matchers.h"
#include "gmock/gmock.h"
#include "google/protobuf/text_format.h"
#include "gtest/gtest.h"
#include "wfa/virtual_people/common/model.pb.h"

namespace wfa_virtual_people {
namespace {

using ::google::protobuf::FieldMask;
using ::wfa::StatusIs;

FieldMask MakeFieldMask(absl::Span<const std::string> paths) {
  FieldMask field_mask;
  for (const std::string& path : paths) {
    field_mask.add_paths(path);
  }
  return field_mask;
}

TEST(HashFieldMaskMatcherTest, EmptyEvents) {
  std::vector<const LabelerEvent*> events;
  FieldMask hash_field_mask = MakeFieldMask({"person_country_code"});
  EXPECT_THAT(HashFieldMaskMatcher::Build(events, hash_field_mask).status(),
              StatusIs(absl::StatusCode::kInvalidArgument, ""));
}

TEST(HashFieldMaskMatcherTest, NullEvent) {
  std::vector<const LabelerEvent*> events;
  events.emplace_back(nullptr);
  FieldMask hash_field_mask = MakeFieldMask({"person_country_code"});
  EXPECT_THAT(HashFieldMaskMatcher::Build(events, hash_field_mask).status(),
              StatusIs(absl::StatusCode::kInvalidArgument, ""));
}

TEST(HashFieldMaskMatcherTest, EmptyHashFieldMask) {
  std::vector<const LabelerEvent*> events;
  LabelerEvent input_event;
  input_event.set_person_country_code("COUNTRY_1");
  events.emplace_back(&input_event);
  FieldMask hash_field_mask;
  EXPECT_THAT(HashFieldMaskMatcher::Build(events, hash_field_mask).status(),
              StatusIs(absl::StatusCode::kInvalidArgument, ""));
}

TEST(HashFieldMaskMatcherTest, EventsHaveSameHash) {
  std::vector<const LabelerEvent*> events;
  LabelerEvent input_event_1;
  input_event_1.set_person_country_code("COUNTRY_1");
  input_event_1.set_person_region_code("REGION_1");
  events.emplace_back(&input_event_1);
  LabelerEvent input_event_2;
  input_event_2.set_person_country_code("COUNTRY_1");
  input_event_2.set_person_region_code("REGION_2");
  events.emplace_back(&input_event_2);

  FieldMask hash_field_mask = MakeFieldMask({"person_country_code"});
  EXPECT_THAT(HashFieldMaskMatcher::Build(events, hash_field_mask).status(),
              StatusIs(absl::StatusCode::kInvalidArgument, ""));
}

TEST(HashFieldMaskMatcherTest, TestGetMatch) {
  std::vector<const LabelerEvent*> events;
  LabelerEvent input_event_1;
  input_event_1.set_person_country_code("COUNTRY_1");
  input_event_1.set_person_region_code("REGION_1");
  events.emplace_back(&input_event_1);
  LabelerEvent input_event_2;
  input_event_2.set_person_country_code("COUNTRY_2");
  input_event_2.set_person_region_code("REGION_2");
  events.emplace_back(&input_event_2);

  FieldMask hash_field_mask = MakeFieldMask({"person_country_code"});
  ASSERT_OK_AND_ASSIGN(std::unique_ptr<HashFieldMaskMatcher> matcher,
                       HashFieldMaskMatcher::Build(events, hash_field_mask));

  // events[0] matches. Returns 0.
  LabelerEvent event_1;
  event_1.set_person_country_code("COUNTRY_1");
  EXPECT_EQ(matcher->GetMatch(event_1), 0);

  // events[0] matches. Returns 0.
  // person_region_code is ignored as not included in hash_field_mask.
  LabelerEvent event_2;
  event_2.set_person_country_code("COUNTRY_1");
  event_2.set_person_region_code("REGION_3");
  EXPECT_EQ(matcher->GetMatch(event_2), 0);

  // events[1] matches. Returns 1.
  LabelerEvent event_3;
  event_3.set_person_country_code("COUNTRY_2");
  EXPECT_EQ(matcher->GetMatch(event_3), 1);

  // events[1] matches. Returns 1.
  // person_region_code is ignored as not included in hash_field_mask.
  LabelerEvent event_4;
  event_4.set_person_country_code("COUNTRY_2");
  event_4.set_person_region_code("REGION_3");
  EXPECT_EQ(matcher->GetMatch(event_4), 1);

  // No matches. Returns -1.
  LabelerEvent event_5;
  event_5.set_person_country_code("COUNTRY_3");
  EXPECT_EQ(matcher->GetMatch(event_5), -1);
}

TEST(HashFieldMaskMatcherTest, TestUnsetField) {
  std::vector<const LabelerEvent*> events;
  LabelerEvent input_event_1;
  input_event_1.set_person_region_code("REGION_1");
  events.emplace_back(&input_event_1);
  LabelerEvent input_event_2;
  input_event_2.set_person_region_code("REGION_2");
  events.emplace_back(&input_event_2);

  FieldMask hash_field_mask =
      MakeFieldMask({"person_country_code", "person_region_code"});
  ASSERT_OK_AND_ASSIGN(std::unique_ptr<HashFieldMaskMatcher> matcher,
                       HashFieldMaskMatcher::Build(events, hash_field_mask));

  // events[0] matches. Returns 0.
  LabelerEvent event_1;
  event_1.set_person_region_code("REGION_1");
  EXPECT_EQ(matcher->GetMatch(event_1), 0);

  // No matches. Returns -1.
  // When using hash field mask, to match an event with unset field, the event
  // being checked must have the same field unset.
  LabelerEvent event_2;
  event_2.set_person_country_code("COUNTRY_1");
  event_2.set_person_region_code("REGION_1");
  EXPECT_EQ(matcher->GetMatch(event_2), -1);
}

}  // namespace
}  // namespace wfa_virtual_people
